src/Microsoft.Identity.Client/Instance/AuthorityEndpointResolutionManager.cs

```diff
@@ -74,34 +72,41 @@ public AuthorityEndpointResolutionManager(IServiceBundle serviceBundle, bool sho
                     MsalError.TenantDiscoveryFailedError,
                     MsalErrorMessage.IssuerWasNotFoundInTheOpenIdConfiguration);
             }
+            
+            var authority = Authority.CreateAuthority(_serviceBundle, authorityInfo.CanonicalAuthority);
+            var tenantId = authority.GetTenantId();
+
+            string authorizationEndpoint = ReplaceTenantToken(edr.AuthorizationEndpoint, tenantId);
+            string tokenEndpoint = ReplaceTenantToken(edr.TokenEndpoint, tenantId);
 
             endpoints = new AuthorityEndpoints(
-                edr.AuthorizationEndpoint.Replace(Constants.Tenant, tenant),
-                edr.TokenEndpoint.Replace(Constants.Tenant, tenant),
-                ReplaceNonTenantSpecificValueWithTenant(edr, tenant));
+                authorizationEndpoint,
+                tokenEndpoint,
+                GetSelfSignedJwtAudience(edr.Issuer, tokenEndpoint, tenantId, authorityInfo.AuthorityType));
 
             Add(authorityInfo, userPrincipalName, endpoints);
             return endpoints;
         }
 
-        internal string ReplaceNonTenantSpecificValueWithTenant(TenantDiscoveryResponse endpoints, string tenant)
+        // Used in WithCertificate to create an audience claim in JWT sent by MSAL to EVO
+        // ADAL uses the token endpoint as audience (tenanted or not)
+        // MSAL had been using the issuer, which does not work when tenantnless. But continue to use issuer for ADFS and B2C.
+        private static string GetSelfSignedJwtAudience(string issuer, string tokenEndpoint, string tenantId, AuthorityType authorityType)

```

Comment:
This is the actual fix. If AAD, use the token endpoint. Else use the issuer. 

Needs more testing with ADFS and B2C to understnad if we can use the token endpoint consistenty.